### ISSUE TITLE
Custom Cow type support

### ISSUE THREAD

**Issue by koivunej** [tafia/quick-protobuf#62]
I've been successfully using quick-protobuf with tokio. However there are a few problems with `tokio-io` codecs which only allow the use of `'static` data. This means I need to convert all parsed structs into owned, which much manual labour or with my new `derive-into-owned`.

Optimal way to parse protobuf messages with tokio would be to allow reading straight from [`BytesMut`](https://docs.rs/bytes/0.4/bytes/struct.BytesMut.html) and to store any values as some kind of wrappers against `BytesMut` (or even downgraded readonly version, does not really matter) or even as `BytesMut` or `Vec` length on-stack versions (both are something like 3 usizes, which would be 12 bytes for 32-bit or even more for 64-bit architectures).

This would however require side-stepping the plain `std::borrow::Cow` a lot. Since there might be even more interesting user specific needs (for example, using some mmap resolved slices) in addition to tokio/bytes usage, there should probably be an option to use user provided Cow-type.

Quickly looking at the generated code, the "custom Cow type" would have to be something like:

 * `fn len(&self) -> usize` for `get_size`
 * `AsRef<[u8]>` for `write_message`
 * `std::default::Default`
 * probably have some `fn read<B: BytesReader>(len: usize, reader: &mut B)` for creating

These could probably all be just duck-typed by specifying the "CowType" as a command line option but BytesReader would have to be enhanced to [split_to](https://docs.rs/bytes/0.4/bytes/struct.BytesMut.html#method.split_to) or at least allow access to the underlying container. It might already have a get_ref or something like that.

Filing this issue here to discuss this more. Would changes supporting this kind of option be welcome?

**Comment by tafia** [comment]
> Optimal way to parse protobuf messages with tokio would be to allow reading straight from BytesMut

I didn't know about bytes crate existence. This looks really useful!

I really welcome such changes even if I am not sure I am the best person to handle it. It looks much lower level than what I'm used to!

**Comment by nerdrew** [comment]
Would you be open to a PR that adds an optional crate feature that enables `Bytes` instead of `Cow<&'a [u8]>` and the same for strings?

**Comment by tafia** [comment]
Definitely!

**Comment by nerdrew** [comment]
I see that support for String and Vec<u8> was merged already. I didn't open a PR for bytes because I thought the performance hit was too large.

**Comment by tafia** [comment]
I think the best is to let people decide if they want performance or convenience, the default being performance.

**Comment by nerdrew** [comment]
Makes sense to me. I'll see if I can rebase and cleanup my `bytes` commit as some point.
